File: src/hearthcms/js/admin-user.js
// ============================================================
// Hearth CMS - Admin User Management
//
// Drives the /admin/users page: a popup dialog for add/edit, plus delete.
// All requests go to /api/admin/users; after any change the table block is
// refreshed from the server-rendered "list" action (the same HTML the page
// was first painted with), so the client never rebuilds rows itself.
//
// Depends on the toast helpers in admin.js (showGoodMessage / showErrorMessage).
// ============================================================
(function () {
    'use strict';

    var API = '/api/admin/users';

    // ----- small helpers -----
    function $(id) { return document.getElementById(id); }

    function post(fields) {
        var fd = new FormData();
        Object.keys(fields).forEach(function (k) { fd.append(k, fields[k]); });
        return fetch(API, { method: 'POST', body: fd }).then(function (r) { return r.json(); });
    }

    function toastOk(t, m) {
        if (typeof showGoodMessage === 'function') showGoodMessage(t, m);
    }
    function toastErr(t, m) {
        if (typeof showErrorMessage === 'function') showErrorMessage(t, m);
        else alert((t ? t + ': ' : '') + (m || ''));
    }

    // ----- dialog open/close -----
    function openDialog() {
        var ov = $('userDialogOverlay');
        if (ov) ov.style.display = 'flex';
        document.body.classList.add('user-dialog-open');
    }

    function closeDialog() {
        var ov = $('userDialogOverlay');
        if (ov) ov.style.display = 'none';
        document.body.classList.remove('user-dialog-open');
    }

    // Reset the form to a clean "create" state.
    function resetForm() {
        $('user_id').value = '0';
        $('userForm').reset();
        $('user_id').value = '0'; // reset() clears hidden too; re-assert
        $('user_role').value = 'admin';
    }

    // ----- create -----
    function openCreate() {
        resetForm();
        $('userDialogTitle').textContent = 'New user';
        $('user_password_label').textContent = 'Password';
        $('user_password').setAttribute('required', 'required');
        $('user_password_hint').textContent = 'Set a password for the new user.';
        openDialog();
        setTimeout(function () { $('user_username').focus(); }, 30);
    }

    // ----- edit -----
    function openEdit(id) {
        post({ action: 'get', id: id }).then(function (d) {
            if (!d.success || !d.data) { toastErr('Load failed', d.message || 'Could not load user.'); return; }
            var u = d.data;
            resetForm();
            $('user_id').value = u.id;
            $('user_username').value = u.username || '';
            $('user_display_name').value = u.display_name || '';
            $('user_email').value = u.email || '';
            $('user_role').value = (u.role === 'editor') ? 'editor' : 'admin';

            $('userDialogTitle').textContent = 'Edit user';
            // Password optional on edit: blank means keep the existing one.
            $('user_password_label').textContent = 'New password';
            $('user_password').removeAttribute('required');
            $('user_password').value = '';
            $('user_password_hint').textContent = 'Leave blank to keep the current password.';

            openDialog();
            setTimeout(function () { $('user_display_name').focus(); }, 30);
        }).catch(function () { toastErr('Network error', 'Please try again.'); });
    }

    // ----- submit (create or update) -----
    function submit(e) {
        if (e) e.preventDefault();
        var form = $('userForm');

        // Native validation (required username, password-on-create, pattern).
        if (form.reportValidity && !form.reportValidity()) return false;

        var fd = new FormData(form);
        fd.append('action', 'save');

        fetch(API, { method: 'POST', body: fd })
            .then(function (r) { return r.json(); })
            .then(function (d) {
                if (d.success) {
                    closeDialog();
                    refreshList();
                    toastOk('Saved', 'User saved.');
                } else {
                    toastErr('Save failed', d.message);
                }
            })
            .catch(function () { toastErr('Network error', 'Please try again.'); });
        return false;
    }

    // ----- delete -----
    function del(id) {
        if (!confirm('Delete this user? This cannot be undone.')) return;
        post({ action: 'delete', id: id }).then(function (d) {
            if (d.success) {
                refreshList();
                toastOk('Deleted', 'User deleted.');
            } else {
                toastErr('Delete failed', d.message);
            }
        }).catch(function () { toastErr('Network error', 'Please try again.'); });
    }

    // ----- refresh the table block from the server -----
    function refreshList() {
        post({ action: 'list' }).then(function (d) {
            if (d.success && d.data && typeof d.data.html === 'string') {
                var wrap = $('userListWrap');
                if (wrap) wrap.innerHTML = d.data.html;
            }
        }).catch(function () { /* leave the stale table; a reload will fix it */ });
    }

    // Backdrop click closes (but clicks inside the dialog don't bubble here).
    function backdrop(e) {
        if (e && e.target && e.target.id === 'userDialogOverlay') closeDialog();
    }

    // Esc closes the dialog.
    document.addEventListener('keydown', function (e) {
        if (e.key === 'Escape') {
            var ov = $('userDialogOverlay');
            if (ov && ov.style.display !== 'none') closeDialog();
        }
    });

    // ----- expose the handlers the inline markup calls -----
    window.userOpenCreate = openCreate;
    window.userOpenEdit = openEdit;
    window.userSubmit = submit;
    window.userDelete = del;
    window.userCloseDialog = closeDialog;
    window.userDialogBackdrop = backdrop;
})();

File: src/hearthcms/engine/RH/AdminUser.cs
using System.Collections.Generic;
using System.Data.SQLite;
using System.Text;
using System.Web;

namespace System.engine.RH
{
    // ============================================================
    // AdminUser - the User Management admin page (/admin/users) and the
    // logic backing its API (/api/admin/users, dispatched by AdminUserApi).
    //
    // The page renders a table of all users. The table body is pre-rendered
    // server-side by AdminUserApi.GetUserListHtml() during the initial page request, so the
    // first paint already shows the list (no extra fetch on load). Add / edit
    // happen in a popup dialog driven by /js/admin-user.js, which posts to the
    // API and then refreshes the table via the same GetUserListHtml() output.
    //
    // Roles are free-text but the UI offers "admin" and "editor". The only
    // hard rule the engine enforces is that the LAST remaining user cannot be
    // deleted (an instance must always keep at least one login), and a user
    // cannot delete their own currently-signed-in account.
    // ============================================================
    public static class AdminUser
    {
        // ===== /admin/users : full page (table pre-populated inline) =====
        public static void HandleRequest()
        {
            if (!AdminGuard.RequireLogin()) return;

            var tpl = new AdminTemplate
            {
                Title = "Users",
                ActiveItem = "users",
                PageHeading = "Users",
                PageHeadingActionsHtml =
                    "<button type='button' class='btn btn-primary btn-sm' onclick='userOpenCreate()'><i class='fa-solid fa-plus'></i> New user</button>",
                ExtraFooterText = "<script src='/js/admin-user.js'></script>\n"
            };

            var sb = new StringBuilder();
            sb.Append(tpl.RenderHeader());

            // Pre-render the list table block on the initial request so the page
            // arrives already populated. The same markup is returned by the API
            // ("list" action) when the client needs to refresh after a change.
            sb.Append(@"
<div id='userListWrap'>
");
            sb.Append(AdminUserApi.GetUserListHtml());
            sb.Append(@"
</div>
");

            // ----- the add/edit dialog (hidden until opened by admin-user.js) -----
            sb.Append(@"
<div id='userDialogOverlay' class='user-dialog-overlay' style='display:none' onclick='userDialogBackdrop(event)'>
    <div class='user-dialog' role='dialog' aria-modal='true' aria-labelledby='userDialogTitle'>
        <div class='user-dialog-head'>
            <h2 id='userDialogTitle'>New user</h2>
            <button type='button' class='icon-btn' onclick='userCloseDialog()' aria-label='Close'><i class='fa-solid fa-xmark'></i></button>
        </div>
        <form id='userForm' onsubmit='return userSubmit(event)' class='user-dialog-body'>
            <input type='hidden' id='user_id' name='id' value='0' />
            <div class='form-field'>
                <label for='user_username'>Username</label>
                <input type='text' id='user_username' name='username' maxlength='80' autocomplete='off' pattern='[a-zA-Z0-9_.-]+' required />
                <small class='form-hint'>Letters, digits, dot, dash and underscore.</small>
            </div>
            <div class='form-field'>
                <label for='user_display_name'>Display name</label>
                <input type='text' id='user_display_name' name='display_name' maxlength='120' autocomplete='off' placeholder='Shown as the author name (optional)' />
            </div>
            <div class='form-field'>
                <label for='user_email'>Email</label>
                <input type='email' id='user_email' name='email' maxlength='200' autocomplete='off' placeholder='(optional)' />
            </div>
            <div class='form-field'>
                <label for='user_role'>Role</label>
                <select id='user_role' name='role'>
                    <option value='admin'>admin</option>
                    <option value='editor'>editor</option>
                </select>
            </div>
            <div class='form-field'>
                <label for='user_password' id='user_password_label'>Password</label>
                <input type='password' id='user_password' name='password' maxlength='200' autocomplete='new-password' />
                <small class='form-hint' id='user_password_hint'>Set a password for the new user.</small>
            </div>
            <div class='user-dialog-actions'>
                <button type='submit' class='btn btn-primary'><i class='fa-solid fa-floppy-disk'></i> Save user</button>
                <button type='button' class='btn btn-ghost' onclick='userCloseDialog()'>Cancel</button>
            </div>
        </form>
    </div>
</div>
");

            sb.Append(tpl.RenderFooter());
            ApiHelper.WriteHtml(sb.ToString());
            ApiHelper.EndResponse();
        }

    }
}
